skglm/solvers/lbfgs.py

```diff
@@ -46,25 +51,46 @@ def _solve(self, X, y, datafit, penalty, w_init=None, Xw_init=None):
             datafit.initialize(X, y)
 
         def objective(w):
-            Xw = X @ w
-            datafit_value = datafit.value(y, w, Xw)
-            penalty_value = penalty.value(w)
+            if self.fit_intercept:
+                Xw = X @ w[:-1] + w[-1]
+                datafit_value = datafit.value(y, w[:-1], Xw)
+                penalty_value = penalty.value(w[:-1])
+            else:
+                Xw = X @ w
+                datafit_value = datafit.value(y, w, Xw)
+                penalty_value = penalty.value(w)
```

Comment:
What do you think about this refactoring @floriankozikowski ?
```suggestion
            w_features = w[:n_features]
            Xw = X @ w_features
            if self.fit_intercept:
                Xw += w[-1]
                
            datafit_value = datafit.value(y, w_features , Xw)
            penalty_value = penalty.value(w_features)
```